protocols/mdns: Use a random alphanumeric string for peer name

jochasinga opened the pull request:
Fixes #2285.

jochasinga committed:
Use a random alphanumeric string for mDNS peer name

Fixes #2285

mxinden commented on protocols/mdns/src/dns.rs:
```diff
@@ -280,14 +282,22 @@ fn segment_peer_id(peer_id: String) -> String {
     // DNS-safe encoding for the Peer ID
     let raw_peer_id = data_encoding::BASE32_DNSCURVE.encode(&peer_id.to_bytes());
     // ensure we don't have any labels over 63 bytes long
     let encoded_peer_id = segment_peer_id(raw_peer_id);
```
With the encoding of peer IDs no longer needed, this and all of the underlying encoding logic can be removed, right?

jochasinga replied:
Yes you're right. Looks like every call to `encode_peer_id()` can also get rid of the `peer_id: &PeerId` argument as well?

jochasinga replied:
Inherently the change to `encode_peer_id()` will affect this public function 
```rust
pub fn build_query_response(
    id: u16,
    peer_id: PeerId, /* unused */
    addresses: impl ExactSizeIterator<Item = Multiaddr>,
    ttl: Duration,
) -> Vec<MdnsPacket>
```
which will no longer need `peer_id` in the argument. Is this change fine?

https://github.com/libp2p/rust-libp2p/blob/6e76435b80209bed1eefd38597b3ba25aef00fa7/protocols/mdns/src/dns.rs#L109

mxinden replied:
:+1: removing `peer_id` from `build_query_response` sounds good to me.

jochasinga replied:
My mistake. Look like `build_query_response` still uses `peer_id` so there's no need to touch it.

mxinden reviewed (commented):
Very cool. Thanks a bunch for picking this up @jochasinga!

Would you mind adding a changelog entry as well?

//CC @peat as you have been working on this logic in the past.

jochasinga committed:
Remove unused code and rename the function

Rename `encode_peer_id` to `generate_peer_id` Because `encode_peer_id`
now does not encode anything.

jochasinga committed:
Update CHANGELOG

Update CHANGELOG and Cargo.toml

mxinden commented on protocols/mdns/src/dns.rs:
```diff
@@ -116,7 +115,7 @@ pub fn build_query_response(
     let addresses = addresses.take(65535);
 
-    let peer_id_bytes = encode_peer_id(&peer_id);
+    let peer_id_bytes = generate_peer_id();
```
```suggestion
    let peer_name_bytes = generate_peer_name();
```

mxinden commented on protocols/mdns/src/dns.rs:
```diff
@@ -260,34 +259,20 @@ fn append_u16(out: &mut Vec<u8>, value: u16) {
-    let encoded_peer_id = segment_peer_id(raw_peer_id);
-    let service_name = str::from_utf8(SERVICE_NAME).expect("SERVICE_NAME is always ASCII");
-    let peer_name = [&encoded_peer_id, service_name].join(".");
+/// Generates a peer ID bytes for a DNS query.
```
```suggestion
/// Generates a random peer name as bytes for a DNS query.
```

mxinden commented on protocols/mdns/src/dns.rs:
```diff
@@ -260,34 +259,20 @@ fn append_u16(out: &mut Vec<u8>, value: u16) {
+    let peer_name = random_string(32 + thread_rng().gen_range(0..32));
 
     // allocate with a little extra padding for QNAME encoding
     let mut peer_id_bytes = Vec::with_capacity(peer_name.len() + 32);
```
```suggestion
    let mut peer_name_bytes = Vec::with_capacity(peer_name.len() + 32);
```

mxinden reviewed (commented):
Looks good to me.

One small nit that just now caught my attention: Given that the peer name is no longer based on the peer ID, would you mind renaming the references from peer ID to peer name?

jochasinga committed:
Rename references from peer ID to peer name

Because the peer name is no longer based on peer ID.

See also: https://github.com/libp2p/rust-libp2p/pull/2311#pullrequestreview-796427041

jochasinga committed:
Ignore the peer name when constructing a response

- Remove the peer ID parsing logic and ignore the peer name.
- Remove `my_peer_id` from `MdnsPeer::new(...)` parameters.
- Ignore addresses with bad or no peer IDs, use the peer ID
of the first address with a peer ID as a reference, and ignore
all addresses whose peer ID does not match the first peer ID.

Note that `MdnsPeer::new(...)` now returns a Result type because
there is no gaurantee that there will be a peer ID to create the peer.

jochasinga commented on protocols/mdns/src/query.rs:
```diff
@@ -218,9 +208,9 @@ impl MdnsPeer {
-        my_peer_id: PeerId,
         ttl: u32,
-    ) -> MdnsPeer {
+    ) -> Result<MdnsPeer, String> {
```
@mxinden @MarcoPolo should this return a `Result<MdnsPeer, _>` or an `Option<MdnsPeer>`? I feel that the first makes it clearer.

MarcoPolo replied:
I think this is a little tricky, but I think it comes down to two things:
1. Is the returned error helpful or actionable? Can it be propagated to the user, or can the caller do something with that error that is useful?
2. Is the returned error correct (This is actually a corollary to the above)

For 1, I don't think the caller can do anything different. Either it can get a peer from the record+packet, or it can't. The error won't change anything. This is true for `MdnsPacket` too. In some other system you can imagine an error leading to a retry, or a user prompt, or some error specific remediation, but not here.

For 2, I don't think the error returned is always correct. There are many branches that return `None` in the logic below, and they aren't all related to a missing peer id. So this error could actually be confusing if the error was actually that there weren't any `TXT` fields (for example).

So my thoughts are that this should be an `Option<...>` like `MdnsPacket`.

jochasinga replied:
Resolved in 7dc6b1f

jochasinga committed:
Returns an Option for MdnsPeer::new(...)

Per @MarcoPolo points that an error isn't useful for the caller
and following MdnsPacket API.

See also: https://github.com/libp2p/rust-libp2p/pull/2311#discussion_r750870918

jochasinga committed:
Clean up with rustfmt

MarcoPolo commented on protocols/mdns/src/query.rs:
```diff
@@ -269,10 +253,13 @@ impl MdnsPeer {
-            addrs,
-            peer_id: my_peer_id,
-            ttl,
+        match my_peer_id {
```
nit: you can simplify this a bit by doing:
```
peer_id.map(|peer_id| MdnsPeer {addrs, peer_id, ttl})
```

jochasinga replied:
Great catch. I never knew that.

MarcoPolo commented on protocols/mdns/src/query.rs:
```diff
@@ -303,3 +290,50 @@ impl fmt::Debug for MdnsPeer {
+                .next()
+                .unwrap();
+
+            let peer = MdnsPeer::new(&packet, record_value, ttl).unwrap();
```
nit, I would make this an `expect` so that if this fails it's a useful message to the test runner.

jochasinga committed:
Minor refactor

Refactor match statement to map since it's more idiomatic and concise.
Use expect in tests for better cue in the test.

mxinden commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+# 0.33.1 [unreleased]
+
+- Use a random alphanumeric string for mDNS peer name (see [PR 2311]).
```
```suggestion
- Use a random alphanumeric string instead of the local peer ID for mDNS peer
  name (see [PR 2311]).

  Note that previous versions of `libp2p-mdns` expect the peer name to be a
  valid peer ID. Thus they will be unable to discover nodes running this new
  version of `libp2p-mdns`.
```

jochasinga replied:
Do I need to correct the `protocols/mdns/CHANGELOG` to 0.34.0 as well here?

mxinden replied:
Yes please :pray:

mxinden commented on protocols/mdns/Cargo.toml:
```diff
@@ -1,7 +1,7 @@
 name = "libp2p-mdns"
 edition = "2018"
-version = "0.33.0"
+version = "0.33.1"
```
```suggestion
version = "0.34.0"
```

While not a breaking change on the API level, it is on the network level. Using a patch release here would make downstream dependents auto-update. Given that this is a breaking change and might split a network, I would prefer it to be an explicit decision to update.

Could you bump the version of `libp2p-mdns` in the root `Cargo.toml` and mention the update in the root `CHANGELOG.md` file as well?

mxinden reviewed (commented):
Code looks good to me. Thank you for the many iterations! 

Just two comments on the release process:

jochasinga committed:
Apply suggestions on release version

Co-authored-by: Max Inden <mail@max-inden.de>

mxinden committed:
*: Fix libp2p-mdns version and changelog entries

mxinden reviewed (approved):
I will merge later today unless anyone objects.

Thanks @jochasinga for all the work and thanks @MarcoPolo for the additional help!

mxinden committed:
Merge branch 'libp2p/master' into 2285-random-peername
